### PR TITLE
Fix issues with read only xml 

### LINKED ISSUE #1470: Additional functionality desired for query_config
#1388 is improving query_config, partly to address #1190 . However, many aspects of https://github.com/ESMCI/cime/issues/1190#issuecomment-282928761 are not addressed in that PR.

I see the most important outstanding requests as:

1. For each compset, a listing of which grids are scientifically and test-supported for that compset

1. Allow subsetting compsets by the first letter of the compset, rather than cam/clm/etc., along with some help text saying "B means XXX, F means YYY", etc. This is important because it's not necessarily intuitive that you get I compsets by specifying `--compsets clm`.

### LINKED ISSUE #2276: query_config no longer works for CESM
query_config no longer works for either cime5.4.0-alpha.23 or on master.
Here is the error for cime5.4.0-alpha.23:

[aliceb@cheyenne1:scripts]>./query_config --components clm
Traceback (most recent call last):
  File "./query_config", line 343, in <module>
    _main_func(__doc__)
  File "./query_config", line 336, in _main_func
    query_component(args.components)
  File "./query_config", line 146, in query_component
    if not all_components and not config_exists:
UnboundLocalError: local variable 'config_exists' referenced before assignment
and here is the error from master:

[aliceb@cheyenne6:scripts]>./query_config --component clm
ERROR: locked

### PR DESCRIPTION
**Pull request by jedwards4b**
Fix issues associated with read-only xml objects, add a test to append config_machines.xml

Test suite: scripts_regression_tests.py
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes #2276  + CDASH

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

File: scripts/lib/CIME/XML/files.py [hunk 1]
```diff
@@ -32,11 +32,15 @@ def __init__(self):
 
     def get_value(self, vid, attribute=None, resolved=True, subgroup=None):
         value = super(Files, self).get_value(vid, attribute=attribute, resolved=False, subgroup=subgroup)
-        if "COMP_ROOT_DIR" not in vid and value is not None and resolved and "COMP_ROOT_DIR" in value:
-            m = re.search("(COMP_ROOT_DIR_[^/]+)/", value)
-            comp_root_dir_var_name = m.group(1)
-            comp_root_dir = self.get_value(comp_root_dir_var_name, attribute=attribute, resolved=False, subgroup=subgroup)
-            self.set_value(comp_root_dir_var_name, comp_root_dir)
+        if value is None and "COMP_ROOT_DIR" not in vid and attribute is not None:
+            value = super(Files, self).get_value(vid, attribute=None, resolved=False, subgroup=subgroup)
```
Comment (jgfouca) [inline scripts/lib/CIME/XML/files.py]: This is a repeat of an earlier line.

**Comment by jgfouca** [comment]
Looks great other than the one comment.

**Comment by rljacob** [comment]
Actual bug issue number is #2276.  I changed the PR description.

File: scripts/lib/CIME/XML/files.py (hunk 1 above)
Comment (jedwards4b) [reply scripts/lib/CIME/XML/files.py]: No it's not, note the change in attribute.

Comment (jgfouca) [reply scripts/lib/CIME/XML/files.py]: Ah, didn't notice that.

**Comment by jedwards4b** [comment]
I've redone this with a simpler, cleaner approach.  Have another look.

**Comment by jedwards4b** [comment]
wait, it's not working yet..

File: scripts/lib/CIME/XML/machines.py [hunk 2]
```diff
@@ -293,6 +293,7 @@ def get_suffix(self, suffix_type):
         return None
 
     def set_value(self, vid, value, subgroup=None, ignore_type=True):
```
[the remark this answers is not in the document]
Comment (jgfouca) [reply scripts/lib/CIME/XML/machines.py]: ok

File: scripts/lib/CIME/XML/files.py [hunk 3]
```diff
@@ -25,12 +25,19 @@ def __init__(self):
...
+        self.COMP_ROOTS = {}
+
         # .config_file.xml at the top level may overwrite COMP_ROOT_DIR_ nodes in config_files
+
         if os.path.isfile(config_files_override):
             self.read(config_files_override)
             self.overwrite_existing_entries()
 
     def get_value(self, vid, attribute=None, resolved=True, subgroup=None):
+        if "COMP_ROOT_DIR" in vid:
```
Comment (jgfouca) [inline scripts/lib/CIME/XML/files.py]: I like this approach.

**Comment by jedwards4b** [comment]
@jgfouca Okay all tests are passing now.

**Comment by kdraeder** [comment]
@jedwards4b 
So it looks like I should go ahead with regression testing in #2277,
after rebasing.

**Pull request merged**